# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

1. This paper is about data tidying and how having tidy datasets makes it easier to analyze and manipulate models. The paper discusses a framework dedicated to creating tools to tidy daya sets.
2. The tidy data standard is intended to provide a clear way to organize data, which then makes the cleaning process much more efficient. The tidy data standard would be used to make sure that it is as easy as possible to analyze the data by making the data organized to begin with so it is easier to work with in the future. The goal is to be able to spend the most time on the anaylsis of data instead of cleaning the data to prepare for the analysis.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [62]:
import pandas as pd
import numpy as np
! git clone https://www.github.com/ToriStoner/assignment2
df=pd.read_csv('./assignment2/data/airbnb_hw.csv')
print(df["Price"])
print(df.tail(20))
pd.to_numeric(df["Price"], errors='coerce')
df["Price"].str.replace(",", "")
df["Price"+'_nan'] = df["Price"].isnull()
print('Total Missings: \n', sum(df["Price"+'_nan']),'\n') #End up with 0 missing values


#2
! git clone https://www.github.com/ToriStoner/assignment2
sharks=pd.read_csv('./assignment2/data/sharks.csv')
print(sharks["Type"])
print(sharks["Type"].unique(), '\n')
sharks["Type"] = sharks["Type"].replace('', np.nan)
print(sharks["Type"].value_counts())
sharks = sharks.rename(columns = {'Boatomg':'Boating'})

#3
! git clone https://www.github.com/ToriStoner/assignment2
pretrial=pd.read_csv('./assignment2/data/pretrial_data.csv')

















fatal: destination path 'assignment2' already exists and is not an empty directory.
0        145
1         37
2         28
3        199
4        549
        ... 
30473    300
30474    125
30475     80
30476     35
30477     80
Name: Price, Length: 30478, dtype: object
        Host Id Host Since                                 Name  \
30458  42874738  8/30/2015               Cozy Room in Manhattan   
30459  42945727  8/30/2015  CLEAN QUEEN SIZE ROOM BEST LOCATION   
30460  42945727  8/30/2015             CLEAN ROOM BEST LOCATION   
30461  42917787  8/30/2015  East Village,  First floor.  new bg   
30462  42952645  8/30/2015  Waterfront Penthouse~5min > TimesSQ   
30463  42942193  8/30/2015    Studio Share Million Dollar Views   
30464  42990630  8/31/2015   Lovely and large Williamsburg Loft   
30465  43008801  8/31/2015  • Sunny Artist Home in Greenpoint •   
30466  42997784  8/31/2015        Beautiful parkfront sanctuary   
30467  42995294  8/31/2015   Colorful, comfortable artsy 2 bd

<ipython-input-62-ef8b26d6d9eb>:15: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  sharks=pd.read_csv('./assignment2/data/sharks.csv')


FileNotFoundError: ignored

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

In [1]:
import pandas as pd
#2
! git clone https://www.github.com/ToriStoner/assignment2
df=pd.read_csv('./assignment2/data/college_completion.csv')
print(df.shape, '\n')
print(df.dtypes, '\n')
print(df.columns[1:5], '\n')
df.head()


#3
pd.crosstab(df['control'], df['level'])


#4
sns.histplot(df['grad_100_value'])
sns.kdeplot(df['grad_100_value'])
sns.boxplot(df['grad_100_value'])
df['grad_100_value'].describe()


#5
var1 = 'grad_100_value'
cat1 = 'control'
cat2 = 'level'
sns.kdeplot(data=df, x=var1, hue=cat1)
sns.kdeplot(data=df, x=var1, hue=cat2)
#need to do groupby and describe, think this should be one plot
var = 'grad_100_value'
group = 'control'
df.loc[:,[group,var]].groupby(group).describe()
var = 'grad_100_value'
group = 'level'
df.loc[:,[group,var]].groupby(group).describe()



#6
df['levelXcontrol']=df['level']+', '+df['control']
var1 = 'levelXcontrol'
cat1 = 'level'
cat2 = 'control'
sns.kdeplot(data = df, x= var1, hue=cat1)
sns.kdeplot(data=df, x=var1, hue=cat2)


#7
var1 = 'aid_value'
cat1 = 'control'
cat2 = 'level'
sns.kdeplot(data=df, x=var1, hue=cat1)
sns.kdeplot(data=df, x=var1, hue=cat2)
var = 'aid_value'
group = 'control'
df.loc[:,[group,var]].groupby(group).describe()
var = 'aid_value'
group = 'level'
df.loc[:,[group,var]].groupby(group).describe()


#8
sns.scatterplot(df['grad_100_value'],df['aid_value'])
var1 = 'grad_100_value'
var2 = 'aid_value'
cat1 = 'control'
cat2 = 'level'
sns.scatterplot(data=df, x=var1, y=var2, hue=cat1, style = cat2)


Cloning into 'assignment2'...
remote: Enumerating objects: 45, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 45 (delta 10), reused 6 (delta 6), pack-reused 31
Receiving objects: 100% (45/45), 5.70 MiB | 19.85 MiB/s, done.
Resolving deltas: 100% (13/13), done.
(3798, 63) 

index                int64
unitid               int64
chronname           object
city                object
state               object
                    ...   
state_sector_ct      int64
carnegie_ct          int64
counted_pct         object
nicknames           object
cohort_size        float64
Length: 63, dtype: object 

Index(['unitid', 'chronname', 'city', 'state'], dtype='object') 



NameError: ignored

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.